sys/net/gnrc/application_layer/tftp/gnrc_tftp.c

```diff
@@ -0,0 +1,973 @@
+        }
+    }
+    else if (m->type != GNRC_NETAPI_MSG_TYPE_RCV) {
+        DEBUG("tftp: unknown message");
+        gnrc_pktbuf_release(outbuf);
+        return BUSY;
+    }
+
+    gnrc_pktsnip_t *pkt = (gnrc_pktsnip_t*)(m->content.ptr);
+
+    assert(pkt->next && pkt->next->type == GNRC_NETTYPE_UDP);
+    assert(pkt->next->next && pkt->next->next->type == GNRC_NETTYPE_IPV6);
+
+    uint8_t *data = (uint8_t*)pkt->data;
+    udp_hdr_t *udp = (udp_hdr_t*)pkt->next->data;
+    ipv6_hdr_t *ip = (ipv6_hdr_t*)pkt->next->next->data;
+
+    xtimer_remove(&(ctxt->timer));
+
+    switch (_tftp_parse_type(data)) {
+    case TO_RRQ:
+    case TO_WRQ: {
+        if (byteorder_ntohs(udp->dst_port) != GNRC_TFTP_DEFAULT_DST_PORT) {
+            /* not a valid start packet */
+            DEBUG("tftp: incoming packet on port %d dropped\n", byteorder_ntohs(udp->dst_port));
+            gnrc_pktbuf_release(outbuf);
+            return FAILED;
+        }
+
+        /* reinitialize the context with the current client */
+        _tftp_init_ctxt(&(ip->src), NULL, ctxt->start_cb, ctxt->data_cb,
+                        _tftp_parse_type(data), ctxt);
+
+        /* get the context of the client */
+        ctxt->dst_port = byteorder_ntohs(udp->src_port);
+        DEBUG("tftp: client's port is %"PRIu16"\n", ctxt->dst_port);
+        int offset = _tftp_decode_start(ctxt, data, outbuf);
+        DEBUG("tftp: offset after decode start = %i\n", offset);
+        if (offset < 0) {
+            DEBUG("tftp: there is no data?\n");
+            gnrc_pktbuf_release(outbuf);
+            return FAILED;
+        }
+
+        /* validate if the application accepts the filename and modes */
+        tftp_action_t action = ctxt->op == TO_RRQ ? TFTP_READ : TFTP_WRITE;
+        if (!ctxt->start_cb(action, ctxt->file_name, ctxt->transfer_size)) {
+            _tftp_send_error(ctxt, outbuf, TE_ACCESS, "Blocked by user application");
+            DEBUG("tftp: callback not able to handle mode\n");
+            return FAILED;
+        }
+
+        /* register a listener for the UDP port */
+        ctxt->entry.next = NULL;
+        ctxt->entry.demux_ctx = ctxt->src_port;
+        ctxt->entry.pid = thread_getpid();
+        gnrc_netreg_register(GNRC_NETTYPE_UDP, &(ctxt->entry));
+
+        /* decode the options */
+        tftp_state state;
+        if (_tftp_decode_options(ctxt, pkt, offset) > offset) {
+            DEBUG("tftp: send option ACK\n");
+            /* the client send the TFTP options, we must OACK */
+            state = _tftp_send_dack(ctxt, outbuf, TO_OACK);
+        }
+        else {
+            DEBUG("tftp: send normal ACK\n");
+            /* the client didn't send options, use ACK and set defaults */
+            _tftp_set_default_options(ctxt);
+
+            /* inform the listener that we are reading */
+            if (ctxt->start_cb) {
+                ctxt->start_cb(TFTP_READ, ctxt->file_name, 0);
+            }
```

Comment:
This prevents the server implementation to work (client put -> server). The start callback will already be called [at line 552](https://github.com/RIOT-OS/RIOT/pull/3911/files#diff-ffc93ed2e621ae21cf7400c802be79e0R552). Calling it again with `TFTP_READ` seems wrong to me.